### PR TITLE
Charlesmchen/background vs contacts permission

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Contacts/SignalRecipient.m [hunk 1]
```diff
@@ -23,7 +23,23 @@ - (instancetype)initWithTextSecureIdentifier:(NSString *)textSecureIdentifier
         return self;
     }
 
-    _devices = [NSMutableOrderedSet orderedSetWithObject:[NSNumber numberWithInt:1]];
+    OWSAssert([TSAccountManager localNumber].length > 0);
+    if ([[TSAccountManager localNumber] isEqualToString:textSecureIdentifier]) {
```
Comment (michaelkirk) [inline SignalServiceKit/src/Contacts/SignalRecipient.m]: This seems unrelated - can you verify this is supposed to be in this PR?

File: Signal/src/contact/SystemContactsFetcher.swift [hunk 2]
```diff
@@ -373,32 +373,40 @@ class SystemContactsFetcher: NSObject {
      *
      * @param   completion  completion handler is called on main thread.
      */
-    public func requestOnce(completion: ((Error?) -> Void)?) {
+    public func requestOnce(completion completionParam: ((Error?) -> Void)?) {
         AssertIsOnMainThread()
 
+        // Ensure completion is invoked on main thread.
+        let completion = { error in
```
Comment (charlesmchen) [inline Signal/src/contact/SystemContactsFetcher.swift]: We weren't always ensuring that the completion was invoked on main.

Comment (charlesmchen) [reply Signal/src/contact/SystemContactsFetcher.swift]: This also avoids the `?()`

File: Signal/src/contact/SystemContactsFetcher.swift [hunk 3]
```diff
@@ -373,32 +373,40 @@ class SystemContactsFetcher: NSObject {
...
         guard !systemContactsHaveBeenRequestedAtLeastOnce else {
-            completion?(nil)
+            completion(nil)
             return
         }
         setupObservationIfNecessary()
 
         switch authorizationStatus {
         case .notDetermined:
+            if UIApplication.shared.applicationState == .background {
```
Comment (charlesmchen) [inline Signal/src/contact/SystemContactsFetcher.swift]: The meat of the matter.

**Comment by charlesmchen** [comment]
PTAL @michaelkirk

File: Signal/src/contact/SystemContactsFetcher.swift (hunk 2 above)
Comment (michaelkirk) [reply Signal/src/contact/SystemContactsFetcher.swift]: > We weren't always ensuring that the completion was invoked on main.
  
Huh, I can't see which case we missed. It looks like they're all invoked on main. 

But I guess with this approach at worst we dispatch an empty block which is NBD.

Comment (charlesmchen) [reply Signal/src/contact/SystemContactsFetcher.swift]: The missed cases are in `updateContacts`, but I decided to rework the completion handling in both methods with the same goal: ensure the that completion is always called on main thread and DRY.

**Comment by charlesmchen** [comment]
> LGTM - please make sure this is in the test plan.

Done.
